### PR TITLE
Cam3 update

### PR DESCRIPTION
**Pull request by vadimzalunin**
CRAM3, indexing and code cleaning

### CODE REVIEW

File: src/java/htsjdk/samtools/CRAMFileReader.java [hunk 1]
```diff
@@ -17,13 +17,14 @@
 
 import htsjdk.samtools.SAMFileHeader.SortOrder;
```
Comment (nh13) [inline src/java/htsjdk/samtools/CRAMFileReader.java]: There is a lot of yellow in intellij related to this file.  For example, on line 123 ```validationStringency``` could be final.  Could you fix those up within this file and others.

File: src/java/htsjdk/samtools/CRAMFileReader.java [hunk 2]
```diff
@@ -244,26 +277,26 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
...
-        try {
-            s.seek(0);
-            si = new CRAMIterator(s, referenceSource);
-            si.setValidationStringency(validationStringency);
-            it = si;
-        } catch (final IOException e) {
-            throw new RuntimeEOFException(e);
-        }
-
         Container c;
```
Comment (nh13) [inline src/java/htsjdk/samtools/CRAMFileReader.java]: Variable names like ```c``` and ```s``` make it extremely hard to follow the code, as we have to keep going back its declaration to know what they are.  Could you please go back through the code and change variable names that reference objects to be more descriptive (ex. ```stream``` or ```container```).

File: src/java/htsjdk/samtools/CRAMFileReader.java [hunk 3]
```diff
@@ -272,6 +305,26 @@ public SAMRecordIterator assertSorted(final SortOrder sortOrder) {
...
+        // Hit the index to determine the chunk boundaries for the required data.
+        if (referenceIndex != -1) {
+            final BAMIndex fileIndex = getIndex();
+            final BAMFileSpan fileSpan = fileIndex.getSpanOverlapping(
+                    referenceIndex, start, -1);
+            filePointers = fileSpan != null ? fileSpan.toCoordinateArray()
+                    : null;
+        }
+
+        if (filePointers == null || filePointers.length == 0)
```
Comment (nh13) [inline src/java/htsjdk/samtools/CRAMFileReader.java]: either have no curly brackets and one line, or curly brackets and multiple lines.

File: src/java/htsjdk/samtools/CRAMFileReader.java [hunk 4]
```diff
@@ -347,4 +394,135 @@ void enableFileSource(final SamReader reader, final boolean enabled) {
...
+            if (overlap) {
+                return CoordMath.overlaps(start, end, interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end);
+            } else {
+                // contained:
+                return CoordMath.encloses(interval.start, interval.end < 0 ? Integer.MAX_VALUE : interval.end, start, end);
+            }
+
+        }
+
+        boolean isPassedTheInterval(SAMRecord record) {
```
Comment (nh13) [inline src/java/htsjdk/samtools/CRAMFileReader.java]: ```isPassedTheInterval``` -> ```isBeyondTheInterval``` or after or past.

**Comment by vadimzalunin** [comment]
fixed all

**Comment by BeatWolf** [comment]
I'm really looking forward to this patch getting merged, as the interval querying support is crucial for my usecase. Any progress on fixing the travis errors?

**Comment by jacarey** [comment]
CRAM 3 support has been rebased to a more current master and all tests have been fixed. Please see the following pull request. I'm closing this one.

https://github.com/samtools/htsjdk/pull/258
